odoo/tools/populate.py

```diff
@@ -1,180 +1,357 @@
+    """
+
+    class Many2oneFieldWrapper(Many2one):
+        def __init__(self, model, field_name, comodel_name):
+            super().__init__(comodel_name)
+            self._setup_attrs(model, field_name)  # setup most of the default attrs
+
+    class Many2manyModelWrapper:
+        def __init__(self, field):
+            self._name = field.relation  # a m2m doesn't have a _name, so we use the tablename
+            self._table = field.relation
+            self._inherits = {}
+            self.env = env
+            self._rec_names_search = []
+            # if the field is inherited, the column attributes are defined on the base_field
+            column1 = field.column1 or field.base_field.column1
+            column2 = field.column2 or field.base_field.column2
+            # column1 refers to the model, while column2 refers to the comodel
+            self._fields = {
+                field.column1: Many2oneFieldWrapper(self, column1, field.model_name),
+                field.column2: Many2oneFieldWrapper(self, column2, field.comodel_name),
+            }
+
+        def __repr__(self):
+            return f"<Many2manyModelWrapper({self._name!r})>"
+
+        def __eq__(self, other):
+            return self._name == other._name
+
+        def __hash__(self):
+            return hash(self._name)
+
+    # check if the relation is an existing model
+    if model := next((model for model in env.registry.models.values() if model._table == field.relation), None):
+        return env[model._name]  # `model` is a MetaModel, re-fetch from env to get the actual ORM model
+    # the relation is a relational table, return a wrapped version
+    return Many2manyModelWrapper(field)
+
+
+def populate_models(env, models, factors, separator_code):
+    """
+    Create factors new records using existing records as templates.
+
+    If a dependency is found for a specific model, but it isn't specified by the user,
+    it will inherit the factor of the dependant model.
+    """
+
+    def has_records(_model):
+        query = SQL('SELECT EXISTS (SELECT 1 FROM %s)', SQL.identifier(_model._table))
+        return _model.env.execute_query(query)[0][0]
+
+    to_process = deque(models)
+    populated = defaultdict(int)  # {model: int(old_max_id)}
+    while to_process:
+        model = to_process.popleft()
+        if not has_records(model):  # if there are no records, there is nothing to populate
+            continue
+        # if the model has _inherits, the delegated models need to have been populated before the current one
+        missing_dependencies = [m for model_name in model._inherits if (m := env[model_name]) not in populated]
+        if missing_dependencies:
+            pending_dependencies = [dep for dep in missing_dependencies if dep not in to_process]
+            if pending_dependencies:
+                to_process.extendleft([model] + pending_dependencies)
```

Comment:
IMHO using a queue for this is inappropriate, since you must process an element's dependencies before the element.  Using a recursive approach would much more readable.  All you need is to know what has been done already (`populated`).  Here is the equivalent of the lines above:
```suggestion
    populated = defaultdict(int)  # {model: int(old_max_id)}
    
    def process(model):
        if model in populated:
            return
        if not has_records(model):  # if there are no records, there is nothing to populate
            populated[model] = 0
            return
        # if the model has _inherits, the delegated models need to have been populated before the current one
        for model_name in model._inherits:
            process(env[model_name])
        # put processing code of model here
```
and the main loop looks like
```py
    for model in models:
        process(model)
```

Reply:
I've applied your suggestion.
The recursion just introduces a slight change of behavior, and it's the sequence that models and their dependent are processed.

Before with the queue, any dependent model (those added via following of a X2many) were processed after *all* CLI models.
This meant that if a dependent model that depends on 2 of the models from the CLI, their Many2one fields would be varied for both of the models, since they were both priorly duplicated.

With a recursion, it's forcing the processing of the dependent models just after the models that added them as a dependent.

TDLR:
If you have models:
`A` --o2m-> `C`
`B` --o2m-> `C`
CLI arg = `A`,`B`
Before the order of population is: `A` > `B` > `C`, then in this case `C` --m2o-> (`A` , `B`) would be transposed to the new copies.
Now with the recursion the order is: `A` > `C` > `B`, so `C` doesn't vary it's --m2o-> `B`, since `B` hasn't been copied yet. 

It's not a *big* difference, just a slight change from before.